src/main/haskell/kore/src/Kore/Step/Simplification/Data.hs

```diff
@@ -55,12 +72,125 @@ simplification of a MetaMLPattern was correct.
 data SimplificationProof level = SimplificationProof
     deriving (Show, Eq)
 
+-- * Branching
+
+{- | 'BranchT' extends any 'Monad' with disjoint branches.
+
+Broadly, one goal of simplification is to promote 'Or' (disjunction) to the top
+level. Many @Simplifier@s return a 'MultiOr' for this reason; we can think of
+this as external branching. @BranchT Simplifier@ also allows branching through
+'Alternative'. Branches are created with '<|>':
+@
+let
+    simplifier1 :: BranchT Simplifier a
+    simplifier2 :: BranchT Simplifier a
+in
+    simplifier1 <|> simplifier2  -- A 'BranchT Simplifier' with two branches.
+@
+
+Branches are pruned with 'empty':
+@
+simplify :: BranchT Simplifier a
+simplify = do
+    unless condition empty
+    continue  -- This simplifier would not be reached if "condition" is 'False'.
+@
+
+Use 'scatter' and 'gather' to translate between the two forms of branches.
+
+-}
+newtype BranchT m a =
+    -- Pay no attention to the ListT behind the curtain!
+    BranchT { runBranchT :: ListT m a }
+    deriving
+        ( Alternative
+        , Applicative
+        , Functor
+        , Monad
+        , MonadIO
+        , MonadPlus
+        , MonadTrans
+        , Typeable
+        )
+
+deriving instance MonadReader r m => MonadReader r (BranchT m)
+
+deriving instance MonadState s m => MonadState s (BranchT m)
+
+{- | Collect the values produced along the disjoint branches.
+ -}
+getBranches :: Applicative m => BranchT m a -> m [a]
+getBranches (BranchT as) = toListM as
+
+{- | Collect results from many simplification branches into one result.
+
+@gather@ collects and merges the results of the internal branches on top and the
+results of the integrated branches below and returns all the results together in
```

Comment:
I think I understand what this function is doing, but I still have a hard time following this comment. To be specific, I can't parse it, I may be missing some context.

Is "on Top" an adjective for "branches", or a place for "collects and merges"? I'm not sure about the entire phrase meaning in any of the two interpretations. What is an integrated branch?

The same applies to "gatherAll".